formats/index/decoder.go

```diff
@@ -0,0 +1,389 @@
+package index
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"io"
+	"io/ioutil"
+	"strconv"
+	"time"
+)
+
+var (
+	// IndexVersionSupported is the range of supported index versions
+	IndexVersionSupported = struct{ Min, Max uint32 }{Min: 2, Max: 4}
+
+	// ErrUnsupportedVersion is returned by Decode when the idxindex file
+	// version is not supported.
+	ErrUnsupportedVersion = errors.New("Unsuported version")
+	// ErrMalformedSignature is returned by Decode when the index header file is
+	// malformed
+	ErrMalformedSignature = errors.New("Malformed index signature file")
+
+	indexSignature          = []byte{'D', 'I', 'R', 'C'}
+	treeExtSignature        = []byte{'T', 'R', 'E', 'E'}
+	resolveUndoExtSignature = []byte{'R', 'E', 'U', 'C'}
+)
+
+type Stage int
+
+const (
+	// Merged is the default stage, fully merged
+	Merged Stage = 1
+	// AncestorMode is the base revision
+	AncestorMode Stage = 1
+	// OurMode is the first tree revision, ours
+	OurMode Stage = 2
+	// TheirMode is the second tree revision, theirs
+	TheirMode Stage = 3
+
+	EntryExtended = 0x4000
+	EntryValid    = 0x8000
+
+	nameMask         = 0xfff
+	intentToAddMask  = 1 << 13
+	skipWorkTreeMask = 1 << 14
+)
+
+type Decoder struct {
+	r         io.Reader
+	lastEntry *Entry
+}
+
+// NewDecoder returns a new decoder that reads from r.
+func NewDecoder(r io.Reader) *Decoder {
+	return &Decoder{r: r}
+}
+
+// Decode reads the whole index object from its input and stores it in the
+// value pointed to by idx.
+func (d *Decoder) Decode(idx *Index) error {
+	version, err := validateHeader(d.r)
+	if err != nil {
+		return err
+	}
+
+	idx.Version = version
+
+	if err := binary.Read(d.r, binary.BigEndian, &idx.EntryCount); err != nil {
+		return err
+	}
+
+	if err := d.readEntries(idx); err != nil {
+		return err
+	}
+
+	return d.readExtensions(idx)
+}
+
+func (d *Decoder) readEntries(idx *Index) error {
+	for i := 0; i < int(idx.EntryCount); i++ {
+		e, err := d.readEntry(idx)
+		if err != nil {
+			return err
+		}
+
+		d.lastEntry = e
+		idx.Entries = append(idx.Entries, *e)
+	}
+
+	return nil
+}
+
+func (d *Decoder) readEntry(idx *Index) (*Entry, error) {
+	e := &Entry{}
+
+	var msec, mnsec, sec, nsec uint32
+
+	flow := []interface{}{
+		&msec, &mnsec,
+		&sec, &nsec,
+		&e.Dev,
+		&e.Inode,
+		&e.Mode,
+		&e.UID,
+		&e.GID,
+		&e.Size,
+		&e.Hash,
+		&e.Flags,
+	}
+
+	if err := readBinary(d.r, flow...); err != nil {
+		return nil, err
+	}
+
+	read := 62
+	e.CreatedAt = time.Unix(int64(msec), int64(mnsec))
+	e.ModifiedAt = time.Unix(int64(sec), int64(nsec))
+	e.Stage = Stage(e.Flags>>12) & 0x3
+
+	if e.Flags&EntryExtended != 0 {
+		read += 2
```

Comment:
Remove this all other magic numbers by making readBinary return the number of bytes read.